Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,782 @@
+#include <XR.h>
+
+#import <UIKit/UIKit.h>
+#import <ARKit/ARKit.h>
+#import <ARKit/ARConfiguration.h>
+#import <MetalKit/MetalKit.h>
+
+@interface SessionDelegate : NSObject <ARSessionDelegate>
+@property (readonly) id<MTLTexture> cameraTextureY;
+@property (readonly) id<MTLTexture> cameraTextureCbCr;
+@end
+
+namespace
+{
+    typedef struct
+    {
+        vector_float2 position;
+        vector_float2 uv;
+        vector_float2 cameraUV;
+    } XRVertex;
+
+    static XRVertex vertices[] =
+    {
+        // 2D positions, UV         camera UV
+        { { -1, -1 },   { 0, 0 },   { 0, 0} },
+        { { -1, 1 },    { 0, 1 },   { 0, 0} },
+        { { 1, -1 },    { 1, 0 },   { 0, 0} },
+        { { 1, 1 },     { 1, 1 },   { 0, 0} },
+    };
+}
+
+/**
+ Implementation of the ARSessionDelegate interface for more info see: https://developer.apple.com/documentation/arkit/arsessiondelegate
+ */
+@implementation SessionDelegate
+{
+    std::vector<xr::System::Session::Frame::View>* activeFrameViews;
+    CVMetalTextureCacheRef textureCache;
```

Comment:
Is everything properly destroyed when we exit XR mode? Did you check to enter/exit XR mode multiple times and see if the memory comsumption is stable?

Reply:
I have not checked this yet, and I suspect it does not. I will check this today.

Reply:
Good idea to check this, stop/resume was actually not working at all!

 I now have that working, and have addressed most of the major memory leaks, however we are still seeing ~15MB (down from 50+ MB!) of leaks between each session.  From looking at the profiler these leaks come from the MtkView's drawables not actually getting released after session end.  As far as I can tell from documentation calling [xrView releaseDrawables] should do the trick, but it doesn't seem to be working.  Do you have any experience working with MtkView is there some other process I should follow to fully deallocate it?

Another interesting thing I have learned is you actually cannot destroy an ARSession you can only pause and then later resume it.  So once we have allocated the AR Session we cannot fully reclaim its resources.